pylint/extensions/typing.py

```diff
@@ -0,0 +1,222 @@
+            return node.args[1].infer(context=context)
+        if inferred.qname() == 'typing.TypeVar':
+            # TypeVar('T2') returns a TypeVar class.
+            return iter([inferred])
+    raise astroid.UseInferenceDefault
+
+def call_looks_like(*names):
+    # type: (*str) -> Callable[[astroid.Call], bool]
+    """Return a function that will tell you if a Call object has a particular name."""
+    s_names = set(names)
+
+    def _looks_like(node):
+        # type: (astroid.Call) -> bool
+        """Return whether a Call object has a particular name."""
+        func = node.func
+        if isinstance(func, astroid.Attribute):
+            return func.attrname in s_names
+        if isinstance(func, astroid.Name):
+            return func.name in s_names
+        return False
+    return _looks_like
+
+def build_assign(value, line_num):
+    # type: (str, int) -> NodeNG
+    """Create an Assign object that can be added to the module.
+
+    Also fixes the line_numbers and the module so context can be shown.
+    """
+    ret = astroid.parse('_ = {}'.format(value)).body[0]
+
+    def _fix_tree(tree):
+        tree.fromlineno = tree.tolineno = tree.lineno = int(line_num)
+        for k in tree.get_children():
+            _fix_tree(k)
+
+    _fix_tree(ret)
+    return ret
+
+TYPE_ANNOTATION = re.compile(r'#\s*type:\s*(.*)')  # # type:
+IGNORE = re.compile(r'\s*ignore\b')
+FUNCTION_ANNOTATION = re.compile(r'(\(.*\))\s*->\s*(.*)')
+STAR_ARGS = re.compile(r'([(,]\s*)\*{1,2}')  # "(*Any" or "... , *Any" or "... , **Any" or
+EMPTY_ARGS = re.compile(r'\(\s*\)')  # ( any-whitespace )
+UNTYPED_ELLIPSIS_ARGS = re.compile(r'\(\s*\.{3}\s*\)')  # ( any-whitespace ... any-whitespace )
+
+def expr_for_comment(comment):
+    # type: (str) -> Optional[str]
+    """Return a compilable string to represent a type annotation."""
+
+    # Skip if the comment does not start with '# type:'.
+    m = TYPE_ANNOTATION.match(comment)
+    if not m:
+        return None
+    expr = m.group(1)  # Everything after '# type:'
+
+    # Skip if it's "# type: ignore"
+    if IGNORE.match(expr):
+        return None
+
+    # Does it look like a signature annotation?
+    m = FUNCTION_ANNOTATION.match(expr)
+    if m:
+        arg_types, return_type = m.groups()
+        if EMPTY_ARGS.match(arg_types) or UNTYPED_ELLIPSIS_ARGS.match(expr):
+            # `...` is not valid Python 2 syntax, so just return the return_type.
+            expr = return_type
+        else:
+            # Remove the parens and any *, ** from the arguments.
+            arg_types = STAR_ARGS.sub(r'\1', arg_types)[1:-1]
+            expr = '%s, %s' % (arg_types, return_type)
+
+    # Skip if expr can't be parsed using compile().
+    try:
+        compile(expr, '<string>', 'eval')
+    except Exception:
+        return None
+
+    return expr
+
+def find_scope(node, lineno):
+    # type: (NodeNG, int) -> Tuple[NodeNG, int]
+    """Find the scope of the code at lineno."""
+    while True:
```

Comment:
This function is very inefficient. I'm changing it to precalculating as much as I can.